src/Path/Elab.hs

```diff
@@ -1,275 +1,206 @@
-    tm ::: ty <- case ty of
-      Just ty -> do
-        scope <- get
-        let ty' = whnf scope ty
-        (names, _) <- un (orTerm (\ n -> \case
-          Value.Pi (Im :< _) b | False -> Just (Im :< n, whnf scope (Value.instantiate (pure n) b))
-          _                    -> Nothing)) ty'
-        pure (Core.lams names tm ::: Value.weaken ty)
-      Nothing -> (tm :::) <$> inferType
-    tm ::: ty <- runScope (define ty (elab tm))
-    modify (Scope.insert name (Entry (Just tm ::: ty)))
-    pure (Define name (tm ::: ty))
-  Doc docs     d  -> Doc docs <$> elabDecl d
-
-declare :: ( Carrier sig m
-           , Effect sig
-           , Member (Error Doc) sig
-           , Member Naming sig
-           , Member (Reader Scope) sig
-           , Member (Reader Span) sig
-           )
-        => ElabC (StateC Signature m) (Value Meta ::: Type Meta)
-        -> m (Value Gensym)
-declare ty = evalState (mempty :: Signature) $ do
-  (constraints, ty') <- runElab (goalIs Type ty)
-  subst <- solver constraints
-  pure (Value.generalizeType (apply subst ty'))
-
-define :: ( Carrier sig m
-          , Effect sig
-          , Member (Error Doc) sig
-          , Member Naming sig
-          , Member (Reader Scope) sig
-          , Member (Reader Span) sig
-          )
-       => Value Meta
-       -> ElabC (StateC Signature m) (Value Meta ::: Type Meta)
-       -> m (Value Gensym ::: Type Gensym)
-define ty tm = evalState (mempty :: Signature) $ do
-  (constraints, tm') <- runElab (goalIs ty tm)
-  subst <- solver constraints
-  let ty' = Value.generalizeType (apply subst ty)
-  (::: ty') <$> Value.strengthen (apply subst tm')
+         => Decl (Term Surface.Surface Qualified)
+         -> m (Decl (Term (Problem :+: Core) Qualified))
+elabDecl (Decl name d tm ty) = namespace (show name) $ do
+  ty' <- runSpanned (goalIs type' . elab) ty
+  def <- meta (unSpanned ty')
+  moduleName <- ask
+  tm' <- runSpanned (local (:> Define (moduleName :.: name := def) ::: unSpanned ty') . goalIs (unSpanned ty') . elab) tm
+  ty'' <- runSpanned (either freeVariables pure . strengthen) ty'
+  tm'' <- runSpanned (either freeVariables pure . strengthen) tm'
+  pure (Decl name d tm'' ty'')
 
-runScope :: (Carrier sig m, Member (State Scope) sig) => ReaderC Scope m a -> m a
-runScope m = get >>= flip runReader m
+elabModule :: ( Carrier sig m
+              , Member (Error Doc) sig
+              , Member Naming sig
+              , Member (Reader (ModuleGraph (Term (Problem :+: Core)) Void)) sig
+              , Member (Writer (Stack Doc)) sig
+              )
+           => Module (Term Surface.Surface) Qualified
+           -> m (Module (Term (Problem :+: Core)) Qualified)
+elabModule m = namespace (show (moduleName m)) . runReader (moduleName m) . local @(ModuleGraph (Term (Problem :+: Core)) Void) (Module.restrict (Map.keysSet (moduleImports m))) $ do
+  -- FIXME: do a topo sort on the decls? or at least make their types known first? or…?
```

Comment:
This is actually not working, so we will definitely need to try something else at some point.